kong/plugins/acl/handler.lua

```diff
@@ -59,47 +62,79 @@ function ACLHandler:access(conf)
     responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
   end
   if not acls then
-    acls = {}
+    acls = EMPTY
   end
 
-  local block
+  -- build and cache a reverse-lookup table from our plugins 'conf' table
+  local reverse = reverse_cache[conf]
+  if not reverse then
+    local groups = {}
+    reverse = {
+      groups = groups,
+      type = (conf.blacklist or EMPTY)[1] and BLACK or WHITE,
+    }
+
+    -- cache by 'conf', the cache has weak keys, so invalidation of the
+    -- plugin 'conf' will also remove it from our local cache here
+    reverse_cache[conf] = reverse
+    -- build reverse tables for quick lookup
+    if reverse.type == BLACK then
+      for i = 1, #(conf.blacklist or EMPTY) do
+        local groupname = conf.blacklist[i]
+        groups[groupname] = groupname
+      end
 
-  if next(conf.blacklist or empty) and next(acls or empty) then
-    for _, v in ipairs(acls) do
-      if utils.table_contains(conf.blacklist, v.group) then
-        block = true
-        break
+    else
+      for i = 1, #(conf.whitelist or EMPTY) do
+        local groupname = conf.whitelist[i]
+        groups[groupname] = groupname
       end
     end
+    -- now create another cache inside this cache for the consumer acls so we
+    -- only ever need to evaluate a white/blacklist once.
+    -- The key for this cache will be 'acls' which will be invalidated upon
+    -- changes. The weak key will make sure our local entry get's GC'ed.
+    -- One exception: a blacklist scenario, and a consumer that does
+    -- not have any groups. In that case 'acls == EMPTY' so all those users
+    -- will be indexed by that table, which is ok, as their result is the
+    -- same as well.
+    reverse.consumer_access = setmetatable({}, { __mode = "k" })
   end
 
-  if next(conf.whitelist or empty) then
-    if not next(acls or empty) then
-      block = true
-    else
-      local contains
-      for _, v in ipairs(acls) do
-        if utils.table_contains(conf.whitelist, v.group) then
-          contains = true
-          break
-        end
+  -- 'cached_result' is either 'true' if it's to be blocked, or the header
+  -- value if it is to be passed
+  local cached_result = reverse.consumer_access[acls]
+  if not cached_result then
+    -- nothing cached, so check our lists and groups
+    local block = (reverse.type == WHITE)
+    for i = 1, #acls do
+      if reverse.groups[acls[i].group] then
+        block = (reverse.type == BLACK)
+        break
       end
-      if not contains then
-        block = true
+    end
+
+    if block then
+      cached_result = true
+
+    else
+      -- allowed, create the header
+      local str_acls = {}
```

Comment:
perf: we can pre-allocate the slots in this table with `new_tab(#acls, 0)` 

And instead of the double O(n) `#acls`, do:

```lua
local n = #acls
local str_acls = new_tab(n, 0)

for i = 1, n do
  -- ..
end
```